ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/altair/util/BlockProcessorAltair.java

```diff
@@ -13,69 +13,151 @@
         validatorsUtil,
         miscHelpers,
-        altairBeaconStateAccessors);
-    this.altairSpecConfig = altairSpecConfig;
-    this.altairBeaconStateAccessors = altairBeaconStateAccessors;
+        beaconStateAccessors,
+        attestationValidator);
+
+    this.specConfigAltair = specConfig;
+    this.miscHelpersAltair = miscHelpers;
+    this.beaconStateAccessorsAltair = beaconStateAccessors;
   }
 
   @Override
-  public void processAttestationsNoValidation(
-      final MutableBeaconState state, final SszList<Attestation> attestations)
-      throws BlockProcessingException {
-    throw new NotImplementedException("TODO");
+  protected void processAttestationNoValidation(
+      final MutableBeaconState genericState, final Attestation attestation) {
+    final MutableBeaconStateAltair state = MutableBeaconStateAltair.required(genericState);
+    final AttestationData data = attestation.getData();
+
+    final SszMutableList<SszByte> epochParticipation;
+    final Checkpoint justifiedCheckpoint;
+    if (data.getTarget().getEpoch().equals(beaconStateAccessors.getCurrentEpoch(state))) {
+      epochParticipation = state.getCurrentEpochParticipation();
+      justifiedCheckpoint = state.getCurrent_justified_checkpoint();
+    } else {
+      epochParticipation = state.getPreviousEpochParticipation();
+      justifiedCheckpoint = state.getPrevious_justified_checkpoint();
+    }
+
+    // Matching roots
+    final boolean isMatchingHead =
+        data.getBeacon_block_root()
+            .equals(beaconStateUtil.getBlockRootAtSlot(state, data.getSlot()));
+    final boolean isMatchingSource = data.getSource().equals(justifiedCheckpoint);
+    final boolean isMatchingTarget =
+        data.getTarget()
+            .getRoot()
+            .equals(beaconStateUtil.getBlockRoot(state, data.getTarget().getEpoch()));
+
+    // Participation flag indices
+    final List<Integer> participationFlagIndices = new ArrayList<>();
+    final UInt64 stateSlot = state.getSlot();
+    final UInt64 dataSlot = data.getSlot();
+    if (isMatchingHead
+        && isMatchingTarget
+        && stateSlot.isLessThanOrEqualTo(
+            dataSlot.plus(specConfig.getMinAttestationInclusionDelay()))) {
+      participationFlagIndices.add(ParticipationFlags.TIMELY_HEAD_FLAG_INDEX);
+    }
+    if (isMatchingSource
+        && stateSlot.isLessThanOrEqualTo(
+            dataSlot.plus(integerSquareRoot(specConfig.getSlotsPerEpoch())))) {
+      participationFlagIndices.add(ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX);
+    }
+    if (isMatchingTarget
+        && stateSlot.isLessThanOrEqualTo(dataSlot.plus(specConfig.getSlotsPerEpoch()))) {
+      participationFlagIndices.add(ParticipationFlags.TIMELY_TARGET_FLAG_INDEX);
+    }
+
+    // Update epoch participation flags
+    UInt64 proposerRewardNumerator = UInt64.ZERO;
+    final List<Integer> attestingIndices =
+        attestationUtil.getAttestingIndices(state, data, attestation.getAggregation_bits());
```

Comment:
It may be unwarranted but I'm slightly nervous about calling this here.  When we validate we wind up creating an `IndexedAttestation` which already has the attesting indices calculated that we could reuse.  Not sure what it would take to pass in here but might be worth logging a follow up ticket at least to check if this is a performance issue.

Reply:
There's an `indexedAttestationCache` used in the `AbstractBlockProcessor` - we could pass that through so we only calculate once.  I'll look into it in a followup 👍

Reply:
Follow-up here: https://github.com/ConsenSys/teku/pull/3792